### PR TITLE
[FLINK-2186] Add readCsvAsRow methods to CsvReader and scala ExecutionEnv

### PR DESCRIPTION
**Pull request by tonycox**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

Rework CSV import to support very wide files

### CODE REVIEW

**Comment by tonycox** [comment]
Hi @StephanEwen could you look at this PR?

**Comment by ex00** [comment]
Thanks for work  @tonycox.
The PR looks good to me. I left a few minor comments about PR.

What do you mean about test negative case? If typeMap does not match with fields type in file for example

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java [hunk 1]
```diff
@@ -192,6 +193,28 @@ public void getFlatFields(String fieldExpression, int offset, List<FlatFieldDesc
...
 
+	public RowTypeInfo(TypeInformation<?> mainType, int size, Map<Integer, TypeInformation<?>> additionalTypes) {
+		this(configureTypes(mainType, size, additionalTypes));
+	}
+
+	public RowTypeInfo(TypeInformation<?> mainType, int size) {
+		this(configureTypes(mainType, size, Collections.<Integer, TypeInformation<?>>emptyMap()));
+	}
+
+	private static TypeInformation<?>[] configureTypes(TypeInformation<?> mainType, int size, Map<Integer, TypeInformation<?>> additionalTypes) {
```
Comment (ex00) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java]: Could you format argumets like as for ```RowTypeInfo#createComparator#219```
for example:
```
private static TypeInformation<?>[] configureTypes(
		TypeInformation<?> mainType,
		int size,
		Map<Integer, TypeInformation<?>> additionalTypes) {
```

File: flink-java/src/main/java/org/apache/flink/api/java/io/CsvReader.java [hunk 2]
```diff
@@ -351,6 +356,32 @@ public CsvReader ignoreInvalidLines(){
 		return new DataSource<T>(executionContext, inputFormat, typeInfo, Utils.getCallLocationName());
 	}
 	
+	public DataSource<Row> rowType(Class<?> mainTargetType, int size, Map<Integer, Class<?>> additionalTypes) {
```
Comment (ex00) [inline flink-java/src/main/java/org/apache/flink/api/java/io/CsvReader.java]: Please add javadoc

File: flink-java/src/main/java/org/apache/flink/api/java/io/CsvReader.java [hunk 3]
```diff
@@ -351,6 +356,32 @@ public CsvReader ignoreInvalidLines(){
...
+			rowTypeInfo = new RowTypeInfo(typeInfo, size);
+		}
+
+		CsvInputFormat<Row> inputFormat = new RowCsvInputFormat(path, rowTypeInfo, this.includedMask);
+		configureInputFormat(inputFormat);
+
+		return new DataSource<Row>(executionContext, inputFormat, rowTypeInfo, Utils.getCallLocationName());
+	}
+
+	public DataSource<Row> rowType(Class<?> mainTargetType, int size) {
```
Comment (ex00) [inline flink-java/src/main/java/org/apache/flink/api/java/io/CsvReader.java]: Please add javadoc

File: flink-scala/src/main/scala/org/apache/flink/api/scala/ExecutionEnvironment.scala [hunk 4]
```diff
@@ -348,6 +349,47 @@ class ExecutionEnvironment(javaEnv: JavaEnv) {
     wrap(new DataSource[T](javaEnv, inputFormat, typeInfo, getCallLocationName()))
   }
 
+  def readCsvFileAsRow[T : ClassTag : TypeInformation](
```
Comment (ex00) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/ExecutionEnvironment.scala]: Could you add scaladoc for method?  may do not understand, what is ```additionalTypes```

File: flink-scala/src/main/scala/org/apache/flink/api/scala/ExecutionEnvironment.scala [hunk 5]
```diff
@@ -348,6 +349,47 @@ class ExecutionEnvironment(javaEnv: JavaEnv) {
...
+    filePath: String,
+    rowSize: Int,
+    additionalTypes: Map[Int, Class[_]] = null,
+    lineDelimiter: String = "\n",
+    fieldDelimiter: String = ",",
+    quoteCharacter: Character = null,
+    ignoreFirstLine: Boolean = false,
+    ignoreComments: String = null,
+    lenient: Boolean = false,
+    includedFields: Array[Int] = null): DataSet[Row] = {
```
Comment (ex00) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/ExecutionEnvironment.scala]: Please add more two spase for parameters like as for other methods in class, space chars should is 6.

File: flink-tests/src/test/java/org/apache/flink/test/io/CsvReaderITCase.java [hunk 6]
```diff
@@ -122,6 +127,80 @@ public void testValueTypes() throws Exception {
...
 
+	private int fullRowSize = 29;
+	private String including = "0111111";
+	private String fileContent =
+		"1,2,3," + 4 + "," + 5.0d + "," + true +
+		",7,8,9,11,22,33,44,55,66,77,88,99,00," +
+		"111,222,333,444,555,666,777,888,999,000\n" +
+		"a,b,c," + 40 + "," + 50.0d + "," + false +
+		",g,h,i,aa,bb,cc,dd,ee,ff,gg,hh,ii,mm," +
+		"aaa,bbb,ccc,ddd,eee,fff,ggg,hhh,iii,mmm\n";
```
Comment (ex00) [inline flink-tests/src/test/java/org/apache/flink/test/io/CsvReaderITCase.java]: A lot of concatenations, the fileContent equals 
```
"1,2,3,4,5.0,true" +
",7,8,9,11,22,33,44,55,66,77,88,99,00," +
"111,222,333,444,555,666,777,888,999,000\n" +
"a,b,c,40,50.0,false," +
"g,h,i,aa,bb,cc,dd,ee,ff,gg,hh,ii,mm," +
"aaa,bbb,ccc,ddd,eee,fff,ggg,hhh,iii,mmm\n"
```

Comment (tonycox) [inline flink-tests/src/test/java/org/apache/flink/test/io/CsvReaderITCase.java]: Yes, but it's simplier to see values, I'll make it static to creating it once.

File: flink-scala/src/main/scala/org/apache/flink/api/scala/ExecutionEnvironment.scala (hunk 5 above)
Comment (tonycox) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/ExecutionEnvironment.scala]: Or 3 tabs :)

**Comment by tonycox** [comment]
@ex00 Thank you for review,
 If typeMap does not match with fields type in file it will fall with `ParseException`.
But if `Double parser` try to parse an integer it will be okay.

Do you have any negative cases that breake this down?

**Comment by ex00** [comment]
Hi @tonycox, thanks for your reply.
>Do you have any negative cases that breake this down?

Yes, i have. which is expected result, if any field in source file is empty, is null or default values or something else?

**Comment by tonycox** [comment]
@ex00 default value of any empty field is null

**Comment by aljoscha** [comment]
I'm closing this as "Abandoned", since there is no more activity and the code base has moved on quite a bit. Please re-open this if you feel otherwise and work should continue.

**Pull request closed without merging**